File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/individuals/client_controller.py
import datetime
from django.contrib import messages
from django.contrib.auth.hashers import make_password
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.shortcuts import get_object_or_404, redirect, render
from django.utils import timezone
from ....Models.individuals.role import Role
from ....Models.individuals.user import User
from ....Models.planning.contract import Contract
from ....Models.planning.estimation import Estimation
from ....Models.planning.plan import Plan
from ....Models.shapes.custom_field import CustomField
from ....Models.utils.utility import Utility
from .._traits.controller import Controller
from .._helpers.http import get_redirect_url
from django.core.exceptions import PermissionDenied
from ....configs.messages_templates import get_exception_class_message

class ClientController(Controller):

  def index(self, request: HttpRequest) -> HttpResponse:
    try:
      if request.user.has_perm('manage client'):
        clients = User.objects.filter(created_by=request.user.creator_id(), type='client')
        return render(request, 'clients/index.html', {'clients': clients})
      else:
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
        return redirect(get_redirect_url(request))
    except Exception as e:
      print(f"[ClientController.index] Failed to load clients: {e.__class__.__name__}: {e}")
      messages.error(request, "ClientController.index: Error")
      return redirect(get_redirect_url(request))

  def create(self, request: HttpRequest) -> HttpResponse:
    try:
      if request.user.has_perm('create client'):
        return (render(request, 'clients/createAjax.html', {})
                if request.GET.get('ajax')
                else render(request, 'clients/create.html', {'customFields': CustomField.objects.filter(module='client')}))
      else:
        return JsonResponse({'error': "Permission Denied."}, status=401)
    except Exception as e:
      print(f"[ClientController.create] Failed to load create view: {e.__class__.__name__}: {e}")
      messages.error(request, "ClientController.create: Error")
      return redirect(get_redirect_url(request))

  def store(self, request: HttpRequest) -> HttpResponse:
    try:
      if request.user.has_perm('create client'):
        default_language_obj = Utility.get_setting_value('default_language', request.user.creator_id())
        name = request.POST.get('name')
        email = request.POST.get('email')
        password = request.POST.get('password')
        if not name or not email or not password:
          error_msg = "Name, Email and Password are required."
          return (JsonResponse({'error': error_msg}, status=401)
                  if request.GET.get('ajax')
                  else (messages.error(request, error_msg) or redirect(get_redirect_url(request))))
        if User.objects.filter(email=email).exists():
          error_msg = "Email already exists."
          return (JsonResponse({'error': error_msg}, status=401)
                  if request.GET.get('ajax')
                  else (messages.error(request, error_msg) or redirect(get_redirect_url(request))))
        objCustomer = request.user
        creator = get_object_or_404(User, pk=objCustomer.creator_id())
        total_client = User.objects.filter(created_by=request.user.creator_id(), type='client').count()
        plan = get_object_or_404(Plan, pk=creator.plan)
        if total_client < plan.max_clients or plan.max_clients == -1:
          role = Role.get_by_name('client')  # TODO: Implement actual role retrieval
          client = User.objects.create(
            name=name,
            email=email,
            job_title=request.POST.get('job_title'),
            password=make_password(password),
            type='client',
            lang=(default_language_obj if default_language_obj else 'en'),
            created_by=request.user.creator_id(),
            email_verified_at=timezone.now(),
          )
          settings_obj = Utility.settings()
          if settings_obj.get('new_client') == 1:
            role_r = Role.get_by_name('client')  # TODO: Replace with actual role assignment
            client.assign_role(role_r)  # TODO: Implement assign_role method on User model
            client.password_plain = password  # Store plain password temporarily for email
            clientArr = {
              'client_name': client.name,
              'client_email': client.email,
              'client_password': client.password_plain,
            }
            resp = Utility.sendEmailTemplate('new_client', [client.email], clientArr)
            success_msg = "Client successfully added." + (
              f"<br> <span class='text-danger'>{resp.get('error')}</span>"
              if resp and not resp.get('is_success') and resp.get('error') else ""
            )
            messages.success(request, success_msg)
            return redirect('clients_index')
          messages.success(request, "Client successfully created.")
          return redirect('clients_index')
        else:
          messages.error(request, "Your user limit is over, Please upgrade plan.")
          return redirect(get_redirect_url(request))
      else:
        return (JsonResponse({'error': "Permission Denied."}, status=401)
                if request.GET.get('ajax')
                else (messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__)) or redirect(get_redirect_url(request))))
    except Exception as e:
      print(f"[ClientController.store] Failed to store client: {e.__class__.__name__}: {e}")
      messages.error(request, "ClientController.store: Error")
      return redirect(get_redirect_url(request))

  def show(self, request: HttpRequest, client_id: int) -> HttpResponse:
    try:
      client = get_object_or_404(User, pk=client_id)
      usr = request.user
      if client and usr.id == client.creator_id() and client.id != usr.id and client.type == 'client':
        estimations = client.clientEstimations().order_by('-id')
        curr_month = client.clientEstimations().filter(issue_date__month=datetime.date.today().month)
        curr_week = client.clientEstimations().filter(
          issue_date__gte=timezone.now().date() - datetime.timedelta(days=timezone.now().date().weekday()),
          issue_date__lte=timezone.now().date() + datetime.timedelta(days=6 - timezone.now().date().weekday())
        )
        last_30days = client.clientEstimations().filter(issue_date__gt=timezone.now().date() - datetime.timedelta(days=30))
        cnt_estimation = {}
        cnt_estimation['total'] = Estimation.getEstimationSummary(estimations)
        cnt_estimation['this_month'] = Estimation.getEstimationSummary(curr_month)
        cnt_estimation['this_week'] = Estimation.getEstimationSummary(curr_week)
        cnt_estimation['last_30days'] = Estimation.getEstimationSummary(last_30days)
        cnt_estimation['cnt_total'] = estimations.count()
        cnt_estimation['cnt_this_month'] = curr_month.count()
        cnt_estimation['cnt_this_week'] = curr_week.count()
        cnt_estimation['cnt_last_30days'] = last_30days.count()
        contracts = client.clientContracts().order_by('-id')
        curr_month_contracts = client.clientContracts().filter(start_date__month=datetime.date.today().month)
        curr_week_contracts = client.clientContracts().filter(
          start_date__gte=timezone.now().date() - datetime.timedelta(days=timezone.now().date().weekday()),
          start_date__lte=timezone.now().date() + datetime.timedelta(days=6 - timezone.now().date().weekday())
        )
        last_30days_contracts = client.clientContracts().filter(start_date__gt=timezone.now().date() - datetime.timedelta(days=30))
        cnt_contract = {}
        cnt_contract['total'] = Contract.getContractSummary(contracts)
        cnt_contract['this_month'] = Contract.getContractSummary(curr_month_contracts)
        cnt_contract['this_week'] = Contract.getContractSummary(curr_week_contracts)
        cnt_contract['last_30days'] = Contract.getContractSummary(last_30days_contracts)
        cnt_contract['cnt_total'] = contracts.count()
        cnt_contract['cnt_this_month'] = curr_month_contracts.count()
        cnt_contract['cnt_this_week'] = curr_week_contracts.count()
        cnt_contract['cnt_last_30days'] = last_30days_contracts.count()
        return render(request, 'clients/show.html', {
          'client': client,
          'estimations': estimations,
          'cnt_estimation': cnt_estimation,
          'contracts': contracts,
          'cnt_contract': cnt_contract
        })
      else:
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
        return redirect(get_redirect_url(request))
    except Exception as e:
      print(f"[ClientController.show] Failed to show client: {e.__class__.__name__}: {e}")
      messages.error(request, "ClientController.show: Error")
      return redirect(get_redirect_url(request))

  def edit(self, request: HttpRequest, client_id: int) -> HttpResponse:
    try:
      if request.user.has_perm('edit client'):
        client = get_object_or_404(User, pk=client_id)
        if client.created_by == request.user.creator_id():
          client.customField = CustomField.getData(client, 'client')  # TODO: Implement getData method
          customFields = CustomField.objects.filter(module='client')
          return render(request, 'clients/edit.html', {'client': client, 'customFields': customFields})
        else:
          return JsonResponse({'error': "Invalid Client."}, status=401)
      else:
        return JsonResponse({'error': "Permission Denied."}, status=401)
    except Exception as e:
      print(f"[ClientController.edit] Failed to load edit view: {e.__class__.__name__}: {e}")
      messages.error(request, "ClientController.edit: Error")
      return redirect(get_redirect_url(request))

  def update(self, request: HttpRequest, client_id: int) -> HttpResponse:
    try:
      if request.user.has_perm('edit client'):
        client = get_object_or_404(User, pk=client_id)
        if client.created_by == request.user.creator_id():
          name = request.POST.get('name')
          email = request.POST.get('email')
          if not name or not email:
            messages.error(request, "Name and Email are required.")
            return redirect(get_redirect_url(request))
          if User.objects.filter(email=email).exclude(pk=client.id).exists():
            messages.error(request, "Email already exists.")
            return redirect(get_redirect_url(request))
          client.name = name
          client.email = email
          if request.POST.get('password'):
            client.password = make_password(request.POST.get('password'))
          client.save()
          CustomField.saveData(client, request.POST.get('customField'))  # TODO: Implement saveData method
          messages.success(request, "Client Updated Successfully!")
          return redirect(get_redirect_url(request))
        else:
          messages.error(request, "Invalid Client.")
          return redirect(get_redirect_url(request))
      else:
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
        return redirect(get_redirect_url(request))
    except Exception as e:
      print(f"[ClientController.update] Failed to update client: {e.__class__.__name__}: {e}")
      messages.error(request, "ClientController.update: Error")
      return redirect(get_redirect_url(request))

  def destroy(self, request: HttpRequest, client_id: int) -> HttpResponse:
    try:
      client = get_object_or_404(User, pk=client_id)
      if client.created_by == request.user.creator_id():
        estimation = Estimation.objects.filter(client_id=client.id).first()
        if not estimation:
          client.delete()
          messages.success(request, "Client Deleted Successfully!")
          return redirect(get_redirect_url(request))
        else:
          messages.error(request, "This client has assigned some estimation.")
          return redirect(get_redirect_url(request))
      else:
        messages.error(request, "Invalid Client.")
        return redirect(get_redirect_url(request))
    except Exception as e:
      print(f"[ClientController.destroy] Failed to delete client: {e.__class__.__name__}: {e}")
      messages.error(request, "ClientController.destroy: Error")
      return redirect(get_redirect_url(request))

  def client_password(self, request: HttpRequest, encrypted_id: str) -> HttpResponse:
    try:
      eId = int(encrypted_id)  # TODO: Replace with proper decryption
    except Exception as e:
      messages.error(request, "Invalid client id.")
      return redirect(get_redirect_url(request))
    user_obj = get_object_or_404(User, pk=eId)
    client = User.objects.filter(created_by=user_obj.creator_id(), type='client').first()
    return render(request, 'clients/reset.html', {'user': user_obj, 'client': client})

  def client_password_reset(self, request: HttpRequest, client_id: int) -> HttpResponse:
    try:
      password = request.POST.get('password')
      password_confirmation = request.POST.get('password_confirmation')
      if not password or password != password_confirmation:
        messages.error(request, "Password confirmation does not match.")
        return redirect(get_redirect_url(request))
      user_obj = get_object_or_404(User, pk=client_id)
      user_obj.password = make_password(password)
      user_obj.save()
      messages.success(request, "Client Password successfully updated.")
      return redirect('clients_index')
    except Exception as e:
      print(f"[ClientController.client_password_reset] Failed to reset password: {e.__class__.__name__}: {e}")
      messages.error(request, "ClientController.client_password_reset: Error")
      return redirect(get_redirect_url(request))

File: _inc/laravel/app/Http/Controllers/Individuals/ClientController.php
<?php

namespace App\Http\Controllers\Individuals;

use App\Http\Controllers\Abstracts\Controller;

use App\Config\Constants\{
    DatabaseConstants as DC,
    MiddlewaresConstants as MWC,
    PermissionsConstants as PMC,
    SettingsConstants as SC,
    UsersConstants as UC,
    ViewsConstants as VW
};
use App\Traits\ChecksLogin;
use App\Models\{
    ClientDeal,
    ClientPermission,
    Contract,
    CustomField,
    Estimation,
    Invoice,
    Plan,
    Role,
    User,
    Utility
};
use App\Services\Reliability\{
    CrmOperationResult,
    CrmOperationService,
    CrmOutboxDispatcher,
    ReliabilityClientPayloadService
};
use Illuminate\Http\{
    JsonResponse,
    RedirectResponse,
    Request
};
use Illuminate\Support\Facades\{Auth, Crypt, DB, Hash, Mail, View as ViewFacade};
use Symfony\Component\HttpFoundation\Response;
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\View\View;

use function App\Http\Controllers\Helpers\defaultPermissionDenial;
use App\Traits\DefinesResourceActions;
class ClientController extends Controller
{
	use DefinesResourceActions;

    use ChecksLogin;
    private const SINGULAR = 'client';

    public function __construct()
    {
        $this->middleware([MWC::AUTH, MWC::XSS]);
    }

    public function index(Request $request): View|RedirectResponse|null
    {
        $action = class_basename(static::class) . '@' . __FUNCTION__;
        $function = __FUNCTION__;
        return $this->measureProfile($action, function () use ($request, $action, $function) {
            try {
                $t = microtime(true);
                self::_authorize($request, PMC::MNG_CLT);
                $this->logExecutionTime($t, $action . '::authorize', 'ok');

                $t = microtime(true);
                $clients = User::where(DC::COL_TABLE_CREATOR, $request->user()->creatorId())
                    ->where(UC::COL_TP, self::SINGULAR)
                    ->get();
                $this->logExecutionTime($t, $action . '::loadClients', 'completed');

                $view = VW::CLT . '.' . $function;
                $t = microtime(true);
                if (!ViewFacade::exists($view)) {
                    $this->logExecutionTime($t, $action . '::viewCheck', 'missing');
                    throw new \RuntimeException("View [$view] not found");
                }
                $this->logExecutionTime($t, $action . '::viewCheck', 'ok');

                return view($view, compact(DC::TABLE_CLIENTS));
            } catch (\Throwable $e) {
                return defaultPermissionDenial($request, $e, $action);
            }
        }, ['uri' => $request->getRequestUri(), 'ip' => $request->ip()]);
    }

    public function create(Request $request): RedirectResponse|View|JsonResponse|null
    {
        $action = class_basename(static::class) . '@' . __FUNCTION__;
        $function = __FUNCTION__;
        return $this->measureProfile($action, function () use ($request, $action, $function) {
            try {
                $t = microtime(true);
                self::_authorize($request, PMC::CR_CLT);
                $this->logExecutionTime($t, $action . '::authorize', 'ok');

                $isAjax = method_exists($request, 'ajax') ? $request->ajax() : ($request->ajax ?? false);

                if ($isAjax) {
                    $view = VW::CLT . '.create_ajax';
                    $t = microtime(true);
                    if (!ViewFacade::exists($view)) {
                        $this->logExecutionTime($t, $action . '::viewCheckAjax', 'missing');
                        throw new \RuntimeException("View [$view] not found");
                    }
                    $this->logExecutionTime($t, $action . '::viewCheckAjax', 'ok');
                    return view($view);
                }

                $t = microtime(true);
                $customFields = CustomField::where('module', self::SINGULAR)->get();
                $this->logExecutionTime($t, $action . '::loadCustomFields', 'completed');

                $view = VW::CLT . '.' . $function;
                $t = microtime(true);
                if (!ViewFacade::exists($view)) {
                    $this->logExecutionTime($t, $action . '::viewCheck', 'missing');
                    throw new \RuntimeException("View [$view] not found");
                }
                $this->logExecutionTime($t, $action . '::viewCheck', 'ok');

                return view($view, compact('customFields'));
            } catch (\Throwable $e) {
                return (method_exists($request, 'ajax') ? $request->ajax() : ($request->ajax ?? false))
                    ? response()->json(['error' => __('Permission Denied.')], 401)
                    : defaultPermissionDenial($request, $e, $action);
            }
        }, ['uri' => $request->getRequestUri()]);
    }

    public function store(Request $request): RedirectResponse|null
    {
        $action = class_basename(static::class) . '@' . __FUNCTION__;
        return $this->measureProfile($action, function () use ($request, $action) {
            try {
                if (($uor = self::_checkLogin()) instanceof \Illuminate\Http\RedirectResponse) {
                    return $uor;
                }
                $creator = $uor;

                $t = microtime(true);
                self::_authorize($request, PMC::CR_CLT);
                $this->logExecutionTime($t, $action . '::authorize', 'ok');

                $t = microtime(true);
                $defaultLang = DB::table(DC::TABLE_SETTINGS)
                    ->where('name', SC::DEF_LNG)
                    ->where(DC::COL_TABLE_CREATOR, $request->user()->creatorId())
                    ->value('value') ?? DC::DEFAULT_LANG;
                $this->logExecutionTime($t, $action . '::loadDefaultLang', 'completed');

                $t = microtime(true);
                $request->validate([
                    'name'     => 'required|string',
                    'email'    => 'required|email|unique:users',
                    'password' => 'required|string',
                ]);
                $this->logExecutionTime($t, $action . '::validate', 'completed');

                $t = microtime(true);
                $plan  = Plan::find($creator->plan());
                $total = User::where(DC::COL_TABLE_CREATOR, $creator->creatorId())
                    ->where(UC::COL_TP, self::SINGULAR)
                    ->count();
                $this->logExecutionTime($t, $action . '::planCheck', 'completed');

                if ($plan->max_clients !== -1 && $total >= $plan->max_clients) {
                    throw new \Exception(__('Your user limit is over. Please upgrade your plan.'));
                }

                $t = microtime(true);
                $operation = (new CrmOperationService())->run(
                    'crm.client.create',
                    function () use ($request, $creator, $defaultLang): array {
                        $client = User::create([
                            'name'              => $request->name,
                            'email'             => $request->email,
                            'job_title'         => $request->job_title,
                            'password'          => Hash::make($request->password),
                            UC::COL_TP => self::SINGULAR,
                            'lang'              => $defaultLang,
                            DC::COL_TABLE_CREATOR => $creator->creatorId(),
                            'email_verified_at' => now()->toDateTimeString(),
                        ]);

                        if ((Utility::settings()['new_client'] ?? 0) == 1) {
                            $client->assignRole(Role::findByName(self::SINGULAR));
                        }

                        return [
                            'client_id' => (string) $client->id,
                            'expected_creator_id' => (string) $creator->creatorId(),
                            'source_table' => DC::TABLE_USERS,
                            'send_email' => (Utility::settings()['new_client'] ?? 0) == 1,
                            'client_name' => (string) $client->name,
                            'client_email' => (string) $client->email,
                            'client_password' => (string) $request->password,
                        ];
                    },
                    [
                        'summary' => 'Create CRM client relationship record',
                        'subject_type' => User::class,
                        'actor_id' => $creator->id,
                        'event_type' => 'crm.client.created',
                        'post_write_validation' => true,
                        'payload' => fn(array $payload): array => $payload,
                    ],
                );
                $this->logExecutionTime($t, $action . '::persistClient', 'completed');

                $resp = [];
                $operationPayload = is_array($operation->value()) ? $operation->value() : [];
                if (($operationPayload['send_email'] ?? false) === true) {
                    $t = microtime(true);
                    $resp = Utility::sendEmailTemplate(
                        'new_client',
                        [(string) $operationPayload['client_email']],
                        [
                            'client_name'     => $operationPayload['client_name'],
                            'client_email'    => $operationPayload['client_email'],
                            'client_password' => $operationPayload['client_password'],
                        ]
                    );
                    $this->logExecutionTime($t, $action . '::sendEmail', 'completed');
                }
                $dispatchReport = $this->dispatchCrmOutbox($operation);

                return redirect()->route(VW::CLT . '.index')->with(
                    'success',
                    __('Client successfully added.')
                        . ((($resp['is_success'] ?? true) === false && ($resp['error'] ?? false))
                            ? '<br><span class="text-danger">' . $resp['error'] . '</span>'
                            : ''
                        )
                )->with('reliability_operation', $this->crmReliabilityPayload($operation, $dispatchReport));
            } catch (AuthorizationException $e) {
                return (method_exists($request, 'ajax') ? $request->ajax() : ($request->ajax ?? false))
                    ? response()->json(['error' => __('Permission Denied.')], 401)
                    : defaultPermissionDenial($request, $e, $action);
            } catch (\Throwable $e) {
                return redirect()->back()->with('error', $e->getMessage());
            }
        }, ['uri' => $request->getRequestUri()]);
    }

    public function show(Request $request, User $client): RedirectResponse|null
    {
        $action = class_basename(static::class) . '@' . __FUNCTION__;
        $function = __FUNCTION__;
        return $this->measureProfile($action, function () use ($request, $client, $action, $function) {
            try {
                $t = microtime(true);
                self::_authorize($request, 'view client');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');

                $t = microtime(true);
                $ownerOk = ($client->creatorId() === $request->user()->id) && ($client->type === self::SINGULAR);
                $this->logExecutionTime($t, $action . '::authorizeOwner', $ownerOk ? 'ok' : 'denied');
                if (!$ownerOk) {
                    throw new \Exception('Invalid Client');
                }

                $t = microtime(true);
                $estimations = $client->clientEstimations()->orderByDesc('id')->get();
                $contracts   = $client->clientContracts()->orderByDesc('id')->get();
                $this->logExecutionTime($t, $action . '::loadRelations', 'completed');

                $summary = fn($col, $getter) => [
                    'total'        => $getter($col),
                    'this_month'   => $getter($col)->whereMonth($col, now()->month),
                    'this_week'    => $getter($col)->whereBetween($col, [now()->startOfWeek(), now()->endOfWeek()]),
                    'last_30days'  => $getter($col)->whereDate($col, '>=', now()->subDays(30)),
                ];

                $t = microtime(true);
                $cntEst = array_map(
                    fn($set) => Estimation::getEstimationSummary($set),
                    $summary('issue_date', fn($c) => $c)
                );
                $cntEst += [
                    'cnt_' . array_key_first($cntEst) => $estimations->count(),
                    'cnt_this_month' => $estimations->whereMonth('issue_date', now()->month)->count(),
                    'cnt_this_week'  => $estimations->whereBetween('issue_date', [now()->startOfWeek(), now()->endOfWeek()])->count(),
                    'cnt_last_30days' => $estimations->whereDate('issue_date', '>=', now()->subDays(30))->count(),
                ];

                $cntCont = array_map(
                    fn($set) => Contract::getContractSummary($set),
                    $summary('start_date', fn($c) => $c)
                );
                $cntCont += [
                    'cnt_total'       => $contracts->count(),
                    'cnt_this_month'  => $contracts->whereMonth('start_date', now()->month)->count(),
                    'cnt_this_week'   => $contracts->whereBetween('start_date', [now()->startOfWeek(), now()->endOfWeek()])->count(),
                    'cnt_last_30days' => $contracts->whereDate('start_date', '>=', now()->subDays(30))->count(),
                ];
                $this->logExecutionTime($t, $action . '::summaries', 'completed');

                $view = VW::CLT . '.' . $function;
                $t = microtime(true);
                if (!ViewFacade::exists($view)) {
                    $this->logExecutionTime($t, $action . '::viewCheck', 'missing');
                    throw new \RuntimeException("View [$view] not found");
                }
                $this->logExecutionTime($t, $action . '::viewCheck', 'ok');

                return view($view, compact(
                    self::SINGULAR,
                    'estimations',
                    'cntEst',
                    DC::TABLE_CONTRACTS,
                    'cntCont'
                ));
            } catch (\Throwable $e) {
                return defaultPermissionDenial($request, $e, $action);
            }
        }, ['client_id' => $client->id, 'uri' => $request->getRequestUri()]);
    }

    public function edit(Request $request, User $client): RedirectResponse|null
    {
        $action = class_basename(static::class) . '@' . __FUNCTION__;
        $function = __FUNCTION__;
        return $this->measureProfile($action, function () use ($request, $client, $action, $function) {
            try {
                $t = microtime(true);
                self::_authorize($request, 'edit client');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');

                $ownerOk = $client->created_by === $request->user()->creatorId();
                $this->logExecutionTime(microtime(true), $action . '::authorizeOwner', $ownerOk ? 'ok' : 'denied');
                if (!$ownerOk) {
                    throw new \Exception('Invalid Client');
                }

                $t = microtime(true);
                $customFields = CustomField::where('module', self::SINGULAR)->get();
                tap($client)->customField = CustomField::getData($client, self::SINGULAR);
                $this->logExecutionTime($t, $action . '::loadFormData', 'completed');

                $view = VW::CLT . '.' . $function;
                $t = microtime(true);
                if (!ViewFacade::exists($view)) {
                    $this->logExecutionTime($t, $action . '::viewCheck', 'missing');
                    throw new \RuntimeException("View [$view] not found");
                }
                $this->logExecutionTime($t, $action . '::viewCheck', 'ok');

                return view($view, [
                    self::SINGULAR => $client,
                    'customFields' => $customFields,
                ]);
            } catch (\Throwable $e) {
                return response()->json(['error' => $e->getMessage()], 401);
            }
        }, ['client_id' => $client->id]);
    }

    public function update(Request $request, User $client): RedirectResponse|null
    {
        $action = class_basename(static::class) . '@' . __FUNCTION__;
        return $this->measureProfile($action, function () use ($request, $client, $action) {
            try {
                $t = microtime(true);
                self::_authorize($request, 'edit client');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');

                $ownerOk = $client->created_by === $request->user()->creatorId();
                $this->logExecutionTime(microtime(true), $action . '::authorizeOwner', $ownerOk ? 'ok' : 'denied');
                if (!$ownerOk) {
                    throw new \Exception('Invalid Client');
                }

                $t = microtime(true);
                $rules = [
                    'name'  => 'required|string',
                    'email' => 'required|email|unique:users,email,' . $client->id,
                ];
                if ($request->filled('password')) {
                    $rules['password'] = 'required|confirmed';
                }
                $request->validate($rules);
                $this->logExecutionTime($t, $action . '::validate', 'completed');

                $t = microtime(true);
                $operation = (new CrmOperationService())->run(
                    'crm.client.update',
                    function () use ($request, $client): array {
                        $client->fill([
                            'name'  => $request->name,
                            'email' => $request->email,
                            'password' => $request->filled('password')
                                ? Hash::make($request->password)
                                : $client->password,
                        ])->save();

                        CustomField::saveData($client, $request->customField);

                        return [
                            'client_id' => (string) $client->id,
                            'expected_creator_id' => (string) $client->created_by,
                            'source_table' => DC::TABLE_USERS,
                        ];
                    },
                    [
                        'summary' => 'Update CRM client relationship record',
                        'subject_type' => User::class,
                        'subject_id' => (string) $client->id,
                        'actor_id' => $request->user()?->id,
                        'event_type' => 'crm.client.updated',
                        'post_write_validation' => true,
                        'payload' => fn(array $payload): array => $payload,
                    ],
                );
                $this->logExecutionTime($t, $action . '::persistClient', 'completed');

                $dispatchReport = $this->dispatchCrmOutbox($operation);

                return redirect()->back()
                    ->with('success', __('Client Updated Successfully!'))
                    ->with('reliability_operation', $this->crmReliabilityPayload($operation, $dispatchReport));
            } catch (\Throwable $e) {
                return redirect()->back()->with('error', $e->getMessage());
            }
        }, ['client_id' => $client->id, 'uri' => $request->getRequestUri()]);
    }

    public function destroy(Request $request, User $client): RedirectResponse|null
    {
        $action = class_basename(static::class) . '@' . __FUNCTION__;
        return $this->measureProfile($action, function () use ($request, $client, $action) {
            try {
                $t = microtime(true);
                self::_authorize($request, 'delete client');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');

                $ownerOk = $client->created_by === $request->user()->creatorId();
                $this->logExecutionTime(microtime(true), $action . '::authorizeOwner', $ownerOk ? 'ok' : 'denied');
                if (!$ownerOk) {
                    throw new \Exception('Invalid Client');
                }

                $t = microtime(true);
                if (Estimation::where('client_id', $client->id)->exists()) {
                    $this->logExecutionTime($t, $action . '::checkDependencies', 'blocked');
                    throw new \Exception(__('This client has assigned some estimation.'));
                }
                $this->logExecutionTime($t, $action . '::checkDependencies', 'ok');

                $t = microtime(true);
                $clientId = (string) $client->id;
                $operation = (new CrmOperationService())->run(
                    'crm.client.delete',
                    function () use ($client, $clientId): array {
                        $client->delete();

                        return [
                            'client_id' => $clientId,
                            'source_table' => DC::TABLE_USERS,
                        ];
                    },
                    [
                        'summary' => 'Delete CRM client relationship record',
                        'subject_type' => User::class,
                        'subject_id' => $clientId,
                        'actor_id' => $request->user()?->id,
                        'event_type' => 'crm.client.deleted',
                        'post_write_validation' => true,
                        'payload' => fn(array $payload): array => $payload,
                    ],
                );
                $dispatchReport = $this->dispatchCrmOutbox($operation);
                $this->logExecutionTime($t, $action . '::delete', 'completed');

                return redirect()->back()
                    ->with('success', __('Client Deleted Successfully!'))
                    ->with('reliability_operation', $this->crmReliabilityPayload($operation, $dispatchReport));
            } catch (\Throwable $e) {
                return redirect()->back()->with('error', $e->getMessage());
            }
        }, ['client_id' => $client->id, 'uri' => $request->getRequestUri()]);
    }

    public const CLT_PSW = 'clientPassword';
    public function clientPassword(string|int $id, Request $request): RedirectResponse|null
    {
        $action = class_basename(static::class) . '@' . __FUNCTION__;
        return $this->measureProfile($action, function () use ($id, $request, $action) {
            if (($uor = self::_checkLogin()) instanceof \Illuminate\Http\RedirectResponse) {
                return $uor;
            }
            $u = $uor;

            try {
                $t = microtime(true);
                self::_authorize($request, 'edit client');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');

                $t = microtime(true);
                $user = User::findOrFail(Crypt::decrypt($id));
                $client = ($u->creatorId() === $user->creatorId() && $user->type == self::SINGULAR)
                    ? $user
                    : throw new \Exception('Invalid Client');
                $this->logExecutionTime($t, $action . '::loadClient', 'completed');

                $view = VW::CLT . '.reset';
                $t = microtime(true);
                if (!ViewFacade::exists($view)) {
                    $this->logExecutionTime($t, $action . '::viewCheck', 'missing');
                    throw new \RuntimeException("View [$view] not found");
                }
                $this->logExecutionTime($t, $action . '::viewCheck', 'ok');

                return view($view, compact('user', self::SINGULAR));
            } catch (\Throwable $e) {
                return redirect()->back()->with('error', $e->getMessage());
            }
        }, ['client_enc_id' => $id, 'uri' => $request->getRequestUri()]);
    }

    public const CLT_PSW_R = 'clientPasswordReset';
    public const IDX = 'index';
    public const CRT = 'create';
    public const STR = 'store';
    public const SHW = 'show';
    public const EDT = 'edit';
    public const UPD = 'update';
    public const DEL = 'destroy';

    public function clientPasswordReset(string|int $id, Request $request): RedirectResponse|null
    {
        $action = class_basename(static::class) . '@' . __FUNCTION__;
        return $this->measureProfile($action, function () use ($id, $request, $action) {
            try {
                $t = microtime(true);
                self::_authorize($request, 'edit client');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');

                $t = microtime(true);
                $request->validate(['password' => 'required|confirmed']);
                $this->logExecutionTime($t, $action . '::validate', 'completed');

                $t = microtime(true);
                User::findOrFail($id)->update(['password' => Hash::make($request->password)]);
                $this->logExecutionTime($t, $action . '::persistPassword', 'completed');

                return redirect()->route(VW::CLT . '.index')
                    ->with('success', 'Client Password successfully updated.');
            } catch (\Throwable $e) {
                return redirect()->back()->with('error', $e->getMessage());
            }
        }, ['client_id' => $id, 'uri' => $request->getRequestUri()]);
    }

    private function dispatchCrmOutbox(CrmOperationResult $operation): ?array
    {
        $message = $operation->outboxMessage();

        return $message ? (new CrmOutboxDispatcher())->dispatchMessage($message) : null;
    }

    /**
     * @param array<string, mixed>|null $dispatchReport
     * @return array<string, mixed>
     */
    private function crmReliabilityPayload(CrmOperationResult $operation, ?array $dispatchReport): array
    {
        return (new ReliabilityClientPayloadService())->fromCrmResult($operation, $dispatchReport);
    }

    protected static function _authorize(
        Request $request,
        string  $permission
    ): void {
        if (!$request->user()->can($permission)) {
            throw new AuthorizationException($permission);
        }
    }
}
